lib/galaxy/webapps/galaxy/services/invocations.py

```diff
@@ -0,0 +1,108 @@
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+)
+
+from pydantic import (
+    BaseModel,
+    Field,
+)
+
+from galaxy.exceptions import AdminRequiredException
+from galaxy.managers.histories import HistoryManager
+from galaxy.managers.workflows import WorkflowsManager
+from galaxy.schema.schema import InvocationIndexQueryPayload
+from galaxy.security.idencoding import IdEncodingHelper
+from galaxy.webapps.galaxy.services.base import ServiceBase
+
+
+class InvocationSerializationParams(BaseModel):
+    """Contains common parameters for customizing model serialization."""
+
+    view: Optional[str] = Field(
+        default=None,
+        title="View",
+        description=(
+            "The name of the view used to serialize this item. "
+            "This will return a predefined set of attributes of the item."
+        ),
+        example="element",
+    )
+    step_details: bool = Field(
+        default=False, title="Include step details", description="Include details for individual invocation steps."
+    )
+    legacy_job_state: bool = Field(
+        default=False,
+        deprecated=True,
+    )
+
+
+class InvocationIndexPayload(InvocationIndexQueryPayload):
+    instance: bool = Field(default=False, description="Is provided workflow id for Workflow instead of StoredWorkflow?")
+
+
+class InvocationsService(ServiceBase):
+    def __init__(
+        self, security: IdEncodingHelper, histories_manager: HistoryManager, workflows_manager: WorkflowsManager
+    ):
+        super().__init__(security=security)
+        self._histories_manager = histories_manager
+        self._workflows_manager = workflows_manager
+
+    def index(
+        self, trans, invocation_payload: InvocationIndexPayload, serialization_params: InvocationSerializationParams
+    ) -> Tuple[List[Dict[str, Any]], int]:
+        workflow_id = invocation_payload.workflow_id
+        if invocation_payload.instance:
+            instance = invocation_payload.instance
+            invocation_payload.workflow_id = self._workflows_manager.get_stored_workflow(
+                trans, workflow_id, by_stored_id=not instance
+            ).id
+        if invocation_payload.history_id:
+            # access check
+            self._histories_manager.get_accessible(
+                invocation_payload.history_id, trans.user, current_history=trans.history
+            )
+        if not trans.user_is_admin:
+            # We restrict the query to the current users' invocations
+            # Endpoint requires user login, so trans.user.id is never None
+            # TODO: user_id should be optional!
+            user_id = trans.user.id
+            if invocation_payload.user_id and invocation_payload.user_id != user_id:
+                raise AdminRequiredException("Only admins can index the invocations of others")
+        else:
+            # Get all invocation if user is admin (and user_id is None)
```

Comment:
I know that's what we've been doing, but each time I look at this I think this is so weird.
```suggestion
            # Get all invocation if user is admin (and user_id is None).
            # TODO: deprecate this behavior, admins should explicitly request that they want to see invocations from all users
```

Reply:
Please add that comment on your own PR instead 😅🤣 - I don't think I totally endorse it? We do the same thing in pages... I think it is... an almost fine behavior. I would prefer it were more consistent - Carl once advocated like an api/admin/<blah> version of the API that would have these behaviors and be a lot more consistent. But we didn't go down that path and I think what we have is totally workable. I will rebase to add that period and fix the typo in invocations here though. I will like this comment also.

Reply:
I mean, you can now write a script that works fine if you're a regular user, but it totally breaks for admins. I don't think that could be considered consistent ?

Reply:
You'd need to test for that. But the endpoint gives you the invocations you have access to right? It is consistent in that respect. "I'm an admin I want to see all the invocations." makes some sense to me. But if you want to break backward compatibility and update the APIs I will back you - it just isn't a priority for me.